replica/table.cc

```diff
@@ -1343,9 +1343,27 @@ future<> table::write_schema_as_cql(database& db, sstring dir) const {
 
 }
 
+// Runs the orchestration code on an arbitrary shard to balance the load.
+future<> table::snapshot_on_all_shards(sharded<database>& sharded_db, const std::vector<foreign_ptr<lw_shared_ptr<table>>>& table_shards, sstring name) {
+    auto jsondir = table_shards[this_shard_id()]->_config.datadir + "/snapshots/" + name;
+    auto orchestrator = std::hash<sstring>()(jsondir) % smp::count;
+
+    co_await smp::submit_to(orchestrator, [&] () -> future<> {
+        auto s = table_shards[this_shard_id()]->schema();
+        tlogger.debug("Taking snapshot of {}.{}: directory={}", s->ks_name(), s->cf_name(), jsondir);
+
+        // FIXME: refactor per-shard snapshot and
+        // run the centralized code here
```

Comment:
It's not a matter of being heavy or not.
The sharded design is problematic since it's fragile, relying on a global snapshot_manager object and aborting if it can't be allocated.
Doing the orchestration here simplifies everything and especially the error case, where we can simply return an error upstream without hanging or aborting.

Reply:
That is not what I meant. I'm asking about the random round-robin load balancing policy that seems to be used here. I saw later that there is quite a lot of code added here, so I guess it doesn't hurt.

Reply:
Yes, describing the schema and preparing the mainfest were already considered heavy enough to be load-balanced.
In general I think we should avoid overloading shard 0 with administrative work (and associated memory consuption), but rather we should try to distribute the load evenly across the shards.

For example, we discussed one another discussion thread having sharded<table> instances.
We shouldn't create and register those only on replica::database shard 0.
A better way would be to use a hash of the table id to select a shard responsible for managing the singleton sharded instance.